fri/src/verifier/mod.rs

```diff
@@ -314,7 +315,16 @@ where
             }
         }
 
-        // make sure the remainder values satisfy the degree
+        // in case `remainder` represents a polynomial of degree `0` then the final check simplifies
+        // to checking that the codeword values are identical. This should probably be combined with
+        // the above so that `remainder[position] != evaluation` is done only once.
+        if remainder.len() <= self.options.blowup_factor(){
+            if !remainder.windows(2).all(|a| a[0] == a[1]) {
+                return Err(VerifierError::RemainderDegreeMismatch(max_degree_plus_1 - 1));
+            }
+        }
+        // make sure the remainder values satisfy the degree bound if the remainder should represent
+        // a polynomial of degree greater than 0.
```

Comment:
I have a couple of questions here (mostly to enhance my understanding as I haven't thought about this too much):

1. Why does remainder length being smaller than blowup factor imply that the remainder should be degree 0 polynomial? Shouldn't this info be in `max_degree_plus_1`? (i.e., if `max_degree_plus_1` is 1, then we are dealing with degree 0 polynomial).
2. Under what combination of parameters (folding factor, blowup factor, degree etc.) would this condition be true?
3. Once we run this check, do we also need to execute `verify_remainder()`? If not, `verify_remainder()` call should probably go into the `else` clause. Or better, yet, I think we can move this check into `verify_remainder()` function. Specifically, in that function we could first check if the remainder should be a degree 0 polynomial. If it is, we run this check and return, if not, we interpolate etc. as we do now.

Reply:
Thank you for the questions:

1. You are right, in this situation `max_degree_plus_1` has all the info we need. I guess my use of the current test is just there from some previous attempts to address some issues that we can discuss somewhere else. As for why the test works, it can be seen that (valid) codewords of length equal to the blowup factor are in fact constant polynomials. Codewords of length less than the blowup factor are, if valid, just the truncation of codewords of length equal to the blowup factor.
2. To answer this, I would like first to remark that `FriOptions` seems to accept for `max_remainder_size` all natural numbers provided they are at least double the size of the folding factor (this I believe is for the Merkle tree construction). This is problematic because of the way the layers are constructed inside the Prover, especially the last layer. For example, taking the folding factor, blowup, max remainder size to be 4, 4 and 8 respectively, will give an error for a trace of length `1<<12` but will work fine for `1<<13`. 
Now to answer the question, let $d^+ := 2^\delta, r := 2^\rho, f:= 2^\phi$ denote the length of the trace, the blowup factor and folding factor, respectively. Suppose `max_remainder_size` as $2^\rho 2^{\delta_m}$ (I think the fact that `max_remainder_size` is bigger than the blowup factor should also be enforced in the code). This implies that the number of layers is given by the smallest $k$ such that: $$\frac{2^\rho 2^\delta}{(2^\phi)^k} \leq 2^\rho 2^{\delta_m}$$.
Simplifying, we get that $$\delta -\delta_m \leq k\phi$$
Now if the left hand side is divisible by $\phi$ cleanly then there are no issues that I can see and the answer to your question will be found by setting $\delta_m$ to $0$. To get this clean division, the left hand side might need to be tweaked and this practically means tweaking $\delta_m$. In our current context we don't want that restriction so we should analyze the second case. If the division is not clean then $k$ is equal to $\lceil\frac{\delta - \delta_m}{\phi}\rceil$. Now for this $k$, the degree of the final codeword will be $\delta_{realized}:= \delta -k\phi$  and everything will work out fine if $\delta_{realized} + \rho > \phi$ so that the final Merkle tree can be constructed.  If we look at the examples given above, we have that $\phi = 2, \rho = 2, \delta_m =  1$, we get that $k=6$ and $\delta_{realized} = 0$ for $\delta = 12$ and so the condition $\delta_{realized} + \rho > \phi$ fails. Thus, to conclude, in some cases we might get errors at the final layer.
Sorry for going on some tangents but I believe that I needed to write this out.
3. You are right, I forgot to put that in an `else` branch. It makes sense to include it inside `verify_remainder()` but don't you think that in this case, since all the elements of the codeword are identical, we can fuse this test with the code block just above it?

Reply:
Thank you @Al-Kindi-0! The answers make sense. Here are my takeaways (and let me know if I misunderstood anything). I'm going to assume that blowup factors is $8$ (as this is what is relevant for us).

1. If we set folding factor to $2$, then we'll get a $0$-degree polynomial in the remainder once remainder gets to exactly $8$ elements. This should be true for any $\delta$.
2. If we set folding factor to $4$, then we'll get the above behavior but only when $\delta$ is even. If $\delta$ is odd, then remainder will contain a degree $1$ polynomial when remainder size gets to $16$.

I think one of the things we should figure out is what are the relative costs of different parameter sets. I think the interesting metrics to track (on per-query basis) would be:
* Number of hashes needed for verifying all Merkle paths.
* Number of hashes needed to hash leaves.
* Complexity of folding procedure.
* Complexity of remainder verification procedure.

I think we could start with the following parameter sets:

* Parameter set 1: field extension degree = $2$, folding factor = $2$.
* Parameter set 2: field extension degree = $3$, folding factor = $2$.
* Parameter set 3: field extension degree = $2$, folding factor = $4$.
* Parameter set 4: field extension degree = $3$, folding factor = $4$.

My initial thinking is that sets 1 and 2 will have a much simpler folding and remainder verification procedures, while parameter sets 3 and 4 will require significantly fewer hashes.

Reply:
Regarding the 3rd point: I'd probably still move into into `verify_remainder()` procedure. I think the performance benefit of doing it the other way is negligible, while the code clarify would be much better if the logic is contained in `verify_remainder()`.

Reply:
>     1. If we set folding factor to 2, then we'll get a 0-degree polynomial in the remainder once remainder gets to exactly 8 elements. This should be true for any δ.

This is correct as then the division is clean trivially by the fact that $\phi = 1$.
> 
>     2. If we set folding factor to 4, then we'll get the above behavior but only when δ is even. If δ is odd, then remainder will contain a degree 1 polynomial when remainder size gets to 16.

What is the value of $\delta_m$ in this case?

Reply:
> I think one of the things we should figure out is what are the relative costs of different parameter sets. I think the interesting metrics to track (on per-query basis) would be:
> 
>     * Number of hashes needed for verifying all Merkle paths.
> 
>     * Number of hashes needed to hash leaves.
> 
>     * Complexity of folding procedure.
> 
>     * Complexity of remainder verification procedure.
> 
> 
> I think we could start with the following parameter sets:
> 
>     * Parameter set 1: field extension degree = 2, folding factor = 2.
> 
>     * Parameter set 2: field extension degree = 3, folding factor = 2.
> 
>     * Parameter set 3: field extension degree = 2, folding factor = 4.
> 
>     * Parameter set 4: field extension degree = 3, folding factor = 4.
> 
> 
> My initial thinking is that sets 1 and 2 will have a much simpler folding and remainder verification procedures, while parameter sets 3 and 4 will require significantly fewer hashes.

I will work on it right away.

Reply:
> Regarding the 3rd point: I'd probably still move into into `verify_remainder()` procedure. I think the performance benefit of doing it the other way is negligible, while the code clarify would be much better if the logic is contained in `verify_remainder()`.

Makes sense!

Reply:
> What is the value of δm in this case?

I think we'd have to set it to 2 so that max remainder length would be 16. The complexity here is that sometimes we'd end with a remainder of length 8 and degree 0, and at other times it would be remainder of length 16 and degree 1. So, the remainder verification procedure would need to handle these two cases.

Btw, verifying that a remainder contains a degree 1 polynomial shouldn't be too complicated, right? Maybe we could do something like provide the slope non-determinisitically, and then use that in the verification process??

Reply:
I think you meant $\delta_m = 1$ because this gives us max remainder length 16.
Indeed, depending on the parity of $\delta - \delta_m$ we would have to handle the two cases you mentioned. I believe that the case of degree one should be easy to do as well, possible even with non-deterministic inputs. I will have to think a bit about it though.